components/camel-milo/src/main/java/org/apache/camel/component/milo/browse/MiloBrowseProducer.java

```diff
@@ -0,0 +1,138 @@
+        super(endpoint);
+
+        this.defaultAwaitWrites = defaultAwaitWrites;
+    }
+
+    @Override
+    public MiloBrowseEndpoint getEndpoint() {
+        return (MiloBrowseEndpoint) super.getEndpoint();
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+        this.connection = getEndpoint().createConnection();
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        if (null != this.connection) {
+            getEndpoint().releaseConnection(connection);
+        }
+        super.doStop();
+    }
+
+    private ExpandedNodeId tryParse(String nodeString) {
+        final Optional<NodeId> nodeId = NodeId.parseSafe(nodeString);
+        return nodeId.map(NodeId::expanded).orElseGet(() -> ExpandedNodeId.parse(nodeString));
+    }
+
+    @Override
+    public boolean process(Exchange exchange, AsyncCallback async) {
+
+        final Message message = exchange.getMessage();
+        final List<ExpandedNodeId> expandedNodeIds = new ArrayList<>();
+
+        if (message.getHeaders().containsKey(HEADER_NODE_IDS)) {
+
+            final List<?> nodes
+                    = message.getHeader(HEADER_NODE_IDS, Collections.singletonList(getEndpoint().getNode()), List.class);
+            message.removeHeader(HEADER_NODE_IDS);
+            if (null == nodes) {
+
+                return false;
+            }
+
+            for (final Object node : nodes) {
+                expandedNodeIds.add(tryParse(node.toString()));
+            }
+        } else {
+
+            expandedNodeIds.add(tryParse(this.getEndpoint().getNode()));
+        }
+
+        final MiloBrowseEndpoint endpoint = this.getEndpoint();
+        final int depth = endpoint.isRecursive() ? endpoint.getDepth() : -1;
+        final boolean subTypes = endpoint.isIncludeSubTypes() || endpoint.isRecursive();
+
+        final CompletableFuture<?> future = this.connection
+                .browse(expandedNodeIds, endpoint.getDirection(), endpoint.getNodeClassMask(), depth, endpoint.getFilter(),
+                        subTypes, endpoint.getMaxNodeIdsPerRequest())
+                .thenApply(browseResults -> {
+
+                    final List<String> expandedNodes = browseResults.values().stream()
+                            .map(BrowseResult::getReferences)
+                            .flatMap(Stream::of)
+                            .map(ReferenceDescription::getNodeId)
+                            .map(ExpandedNodeId::toParseableString)
+                            .collect(Collectors.toList());
+
+                    // For convenience, to be used with the milo-client producer
+                    exchange.getMessage().setHeader(HEADER_NODE_IDS, expandedNodes);
+
+                    exchange.getMessage().setBody(browseResults);
+
+                    return browseResults;
+                });
+
+        final Boolean await = message.getHeader("await", this.defaultAwaitWrites, Boolean.class);
```

Comment:
This seems wrong that a header determines to wait for the future - what is the special use-case. We do not do this in other components, so please remove.

Reply:
Sorry, that's another left-over from the milo-client component, which I used as a template for the milo-browse component. I removed that in milo-browse, as that indeed doesn't make sense in case of the browsing functionality. I guess the intention of the original author of the milo-client might be to fake synchronous execution for writes, even though they are in fact executed asynchrously (in case of defaultAwaitWrites=false). Just my two cents...